bottombar.py

```diff
@@ -20,152 +20,303 @@
+
+    def __call__(self):
+        while self.__stack:
+            f, args = self.__stack.pop()
+            f(*args)
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, *args):
+        self()
+
+
+try:
+    from signal import signal, setitimer, getitimer, SIGWINCH, SIGALRM, ITIMER_REAL
+except ImportError:
+    from _thread import allocate_lock, start_new_thread
+
+    class Auto:
+        '''Register handler function for resize and time events.
+
+        This general implementation creates a thread upon first registration
+        from which the handler is called at a given refresh rate, and the
+        screen size is polled every second. The thread is closed when the
+        original handler is restored.'''
+
+        def __init__(self, handler):
+            self.handler = handler
+            self.refresh = None
+            self.lock = allocate_lock()
+            self.lock.acquire()
+
+        def __call__(self, refresh=float('inf')):
+            restore = CallStack()
+            if not self.refresh:
+                restore[self.__set] = self.refresh,
+                self.refresh = refresh
+                if self.lock.locked():
+                    start_new_thread(self.__run, ())
+            elif refresh < self.refresh:
+                restore[self.__set] = self.refresh,
+                self.__set(refresh)
+            return restore
+
+        def __set(self, refresh):
+            self.refresh = refresh
+            if self.lock.locked():
+                self.lock.release()
+
+        def __run(self):
+            timeout = self.refresh
+            poll_rate = 1.
+            if poll_rate < timeout:
+                size = os.get_terminal_size()
+            while timeout:
+                if not self.lock.acquire(timeout=min(timeout, poll_rate)):
+                    # timed out
+                    if poll_rate < timeout:
+                        old_size = size
+                        size = os.get_terminal_size()
+                        if size == old_size:
+                            timeout -= poll_rate
+                            continue
+                        # size changed
+                    self.handler()
+                timeout = self.refresh
+
+else: # Unix
+
+    class Auto:
+        '''Register handler function for resize and time events.
+
+        This Unix-specific implementation relies on the operating system's
+        signal mechanism to watch for SIGWINCH and SIGALRM events. As a result,
+        pre-existing handlers for those two events are temporarily disabled.'''
+
+        def __init__(self, handler):
+            self.handler = lambda sig, frame: handler()
+            self.refresh = None
+
+        def __call__(self, refresh=float('inf')):
+            restore = CallStack()
+            restore[setattr] = self, 'refresh', self.refresh
+            if not self.refresh:
+                restore[signal] = SIGWINCH, signal(SIGWINCH, self.handler)
```

Comment:
Forward signals to the original handler.

Reply:
done